### PR TITLE
Allow templateinstance controller to instantiate non-v1 objects

### LINKED ISSUE #14780: Template broker can't provision a template that contains a Deployment resource
Using a template like this:

```
apiVersion: v1
kind: Template
metadata:
  name: deployment-simple-test
objects:
- apiVersion: apps/v1beta1
  kind: Deployment
  metadata:
    name: hello-openshift
  spec:
    replicas: 1
    template:
      metadata:
        labels:
          app: nginx
      spec:
        containers:
        - name: hello-openshift
          image: openshift/hello-openshift
          ports:
          - containerPort: 80
```

The error being returned by the service catalog on the service instance is:

```
Provision call failed: Error deprovisioning Instance "temptest6/deployment-simple-test-gnld6" of ServiceClass "deployment-simple-test" at Broker "template-broker": ""
```

### PR DESCRIPTION
**Pull request by jim-minter**
fixes #14780

### CODE REVIEW

**Comment by jim-minter** [comment]
- [ ] implement test case

**Comment by jim-minter** [comment]
@deads2k @soltysh please can you take a look at this?  It seems to solve the issue raised, but:

1) I've copied code from https://github.com/openshift/origin/blob/master/pkg/build/admission/jenkinsbootstrapper/admission.go#L107-L133 , which is in itself naughty
2) https://github.com/openshift/origin/blob/master/pkg/project/registry/projectrequest/delegated/delegated.go#L174 is also doing this wrong

I think a helper function like https://github.com/openshift/origin/blob/master/pkg/cmd/util/clientcmd/factory.go#L105 would be nice, but I don't know how to create all the clientcmd scaffolding in the contexts above (or better, how to avoid creating all the clientcmd scaffolding).

Please can you accept or suggest an acceptable achievable refactor?

@bparees

**Comment by soltysh** [comment]
I'm fine with this copy&paste for now. We'll probably want to address that in 3.7, properly. The only missing thing here is a test-case proving that different k8s resources can be easily created. @deads2k agreed?

File: pkg/template/controller/templateinstance_controller.go [hunk 1]
```diff
@@ -307,10 +311,35 @@ func (c *TemplateInstanceController) instantiate(templateInstance *templateapi.T
 			RESTMapper:  client.DefaultMultiRESTMapper(),
 			ObjectTyper: kapi.Scheme,
 			ClientMapper: resource.ClientMapperFunc(func(mapping *meta.RESTMapping) (resource.RESTClient, error) {
+				config := *c.config
+
+				// TODO copied from pkg/cmd/util/clientcmd/factory_object_mapping.go#ClientForMapping
```
Comment (deads2k) [inline pkg/template/controller/templateinstance_controller.go]: Why can't you move this to an openshift `/resource` package that returns a clientmapperfunc based on a config?

**Comment by deads2k** [comment]
> Please can you accept or suggest an acceptable achievable refactor?

Suggestion made.  I think that centralizing the problem and creating a clientmapperfunc from a config would be sufficient to move us forward.

**Comment by jim-minter** [comment]
updated.

[test][testextended][extended:core(templates)]

**Comment by bparees** [comment]
[merge][severity:blocker]

**Comment by bparees** [comment]
removed the merge tag, this appears to have panicked our extended tests:
https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended/691/testReport/junit/(root)/Extended/BeforeSuite/

haven't looked into it to confirm it is the culprit.

**Comment by jim-minter** [comment]
oops, too many threads in parallel - fixed and repushed

**Comment by bparees** [comment]
[merge][severity:blocker]

**Comment by bparees** [comment]
```
@@ -7319,7 +7319,7 @@ items:
         metadata:
           name: secret
       - kind: Deployment
-        apiVersion: extensions/v1beta1
+        apiVersion: apps/v1beta1
         metadata:
           name: deployment
         spec:
FAILURE: Generated test/extended bindata out of date. Please run hack/update-generated-bindata.sh
```

**Comment by jim-minter** [comment]
fixed and repushed

**Comment by soltysh** [comment]
LGTM

**Comment by bparees** [comment]
[merge][severity:blocker]

**Pull request merged**